### ISSUE TITLE
Font rendering problem: Lines below characters on dark backgrounds

### ISSUE THREAD

**Issue by dnlmlr** [QuantumBadger/Speedy2D#32]
# The problem
When drawing text on dark backgrounds, there are visible Lines below the text in the same color as the text.
This happens even in the hello world example when changing the background color to black. The examples use the `Karla` font, but the same thing happened with all other fonts as well.

## Examples:
![grafik](https://user-images.githubusercontent.com/34707428/129889272-0e3a5e87-559e-404d-a318-393ca501e2f5.png)
![grafik](https://user-images.githubusercontent.com/34707428/129889316-7e977f43-50d2-44fc-81c8-32e5f1e36407.png)

# Steps to reproduce
1. Get the hello world example
2. Set the clear color to BLACK and the text color to WHITE
3. Run the program

# Environment
- OS: Windows
- Speedy2d version: 1.1.0
- Rust version: 1.52.1

**Comment by dnlmlr** [comment]
It seems that the line is always on the bottom most pixel of the glyph bounding box. For characters that "go lower", the line is on the bottom as well.
![grafik](https://user-images.githubusercontent.com/34707428/129916524-ce1c8d9c-a550-457c-9295-02a178fc7dc4.png)

**Comment by dnlmlr** [comment]
So the problem definitely does not come from `rusttype` and it most likely has to do with a bounding box issue on the texture mapping. I modified the `draw_glyph` function from [`font_cache.rs`](https://github.com/QuantumBadger/Speedy2D/blob/2bccdfcf79d35d03e296285015149f9817f7f559/src/font_cache.rs#L495) to force every font pixel to be black, but the lines still remain visible while the actual characters are blacked out. Also it seems that the glyph textures are filled at the positions outside the glyph bounding box, which would explain the line. If the rectange that is taken from the texture is 1 px too high, it would include a line on the bottom that comes from the background filling.
 
```rust
    fn draw_glyph(&mut self, glyph: &rusttype::PositionedGlyph)
    {
        glyph.draw(|x, y, alpha| {
            let start = (4 * (self.size.x * y + x)) as usize;
            self.data[start] = 0;
            self.data[start + 1] = 0;
            self.data[start + 2] = 0;
            self.data[start + 3] = (alpha * 255.0).round() as u8;
        })
    }
```

**Comment by QuantumBadger** [comment]
Thanks for the report! I can reproduce this on Windows but not on Linux -- I think that may be due to multisampling, as this is disabled on the Linux machines I've tried this on, but enabled on Windows.

Possibly an issue with the texture coordinates being misaligned somehow?

**Comment by dnlmlr** [comment]
It seems to be indeed a problem with multisampling. When creating the window with `WindowCreationOptions` that have `with_multisampling(0)`, the lines do not show up. I noticed that the textures used for the glyph cache are filled with `255`, opaque white by default and only the area around the actual rasterized glyphs is made transparent. 
So the line on the bottom might be comming from the MSAA taking the white pixels outside of the intended glyph area into consideration. I'm not quite sure why it would do that since the texture coordinates seem to line up with their intended values perfectly.

So I don't know what the root cause of this issue is yet, nor can I offer a real solution. What I can offer though is a fix: If the glyph texture is set to transparent by default, there is no white filling that could show up. Two options to do that would be the following:

1. In `BitmapRGBA::new` set the whole data vector to `0` instead of `255` and in `BitmapRGBA::draw_glyph` set the color values to `255` in addition to setting the alpha channel.
```rust
    fn draw_glyph(&mut self, glyph: &rusttype::PositionedGlyph)
    {
        glyph.draw(|x, y, alpha| {
            let start = (4 * (self.size.x * y + x)) as usize;
            self.data[start + 3] = (alpha * 255.0).round() as u8;

            self.data[start] = 255;
            self.data[start + 1] = 255;
            self.data[start + 2] = 255;
        })
    }
```

2. The other option would be to just set the alpha channel of each pixel to `0` in `GlyphCacheTexture::new`. When the glyph is rendered, the alpha channels get overwritten, so everything just works. (The iterator is most likely not the best performing way to do this, but this is just indended as an example)
```rust
fn new(context: &GLContextManager) -> Result<Self, BacktraceError<ErrorMessage>>
    {
        let mut bitmap = BitmapRGBA::new(Vector2::new(
            GlyphCacheTexture::SIZE,
            GlyphCacheTexture::SIZE
        ));

        bitmap.data.iter_mut().skip(3).step_by(4).for_each(|it| *it = 0);

        Ok(GlyphCacheTexture {
            bitmap,

            texture: context
                .new_texture()
                .context("GPU texture creation failed")?,

            invalidated: false,

            packer: TexturePacker::new(GlyphCacheTexture::SIZE, GlyphCacheTexture::SIZE),

            entries: HashMap::new()
        })
    }
```

Both of those options mitigate the problem. Let me know what you think

**Comment by QuantumBadger** [comment]
Thanks! I think the best solution for this would be to add a one-pixel transparent border around each glyph in the cache texture. That way there should be no interference from the default background, or from other glyphs (which may go right up to the edge of their allocated space).

If you'd be interested in submitting a PR for this I'd happily accept :) Otherwise I'll probably get around to this within the next couple of weeks.

**Comment by QuantumBadger** [comment]
I've resolved this issue in https://github.com/QuantumBadger/Speedy2D/commit/81679866f7448ea8276412995f98c2cda1e2dc7a, and I've just published v1.1.2 with the fix in. Please let me know if it works for you :)

There seems to be an unrelated issue with dark text on a dark background having a light border when antialiasing is turned off. I'll open up a separate issue for this.
